### PR TITLE
feat: Add support for environment variable FIRESTORE_PREFER_REST

### PR DESCRIPTION
**Pull request by MarkDuckworth**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/nodejs-firestore/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: dev/test/util.ts [hunk 1]
```diff
@@ -33,4 +34,80 @@ describe('isPlainObject()', () => {
...
+    });
+
+    it('ignores case', async () => {
+      process.env.FIRESTORE_PREFER_REST = 'True';
+      const [prevSet, prevValue] = tryGetPreferRestEnvironmentVariable();
+      expect(prevSet).to.be.true;
+      expect(prevValue).to.be.true;
+    });
+
+    it('indicates when the environment variable is not set', async () => {
```
Comment (dconeybe) [inline dev/test/util.ts:94]: "indicates" maybe isn't the right word here. I interpreted it to mean that the test is verifying that there is some "indication" that the environment variable is not set; however, there is no such "indication".

File: dev/test/util.ts [hunk 2]
```diff
@@ -33,4 +34,80 @@ describe('isPlainObject()', () => {
...
+    });
+
+    it('reads 0', async () => {
+      process.env.FIRESTORE_PREFER_REST = '0';
+      const [prevSet, prevValue] = tryGetPreferRestEnvironmentVariable();
+      expect(prevSet).to.be.true;
+      expect(prevValue).to.be.false;
+    });
+
+    it('ignores case', async () => {
```
Comment (dconeybe) [inline dev/test/util.ts:87]: Add a test for "strips leading and trailing whitespace"

**Review by dconeybe — CHANGES_REQUESTED** [review]
Overall looks fantastic! Great test coverage and well designed. Thanks for making this improvement.

File: dev/src/index.ts [hunk 3]
```diff
@@ -534,7 +535,10 @@ export class Firestore implements firestore.Firestore {
    * for communication from that point forward. Currently the only operation
    * that requires gRPC is creating a snapshot listener with the method
    * `DocumentReference<T>.onSnapshot()`, `CollectionReference<T>.onSnapshot()`, or
-   * `Query<T>.onSnapshot()`.
+   * `Query<T>.onSnapshot()`. If specified, this setting value will take precedent over the
+   * environment variable `FIRESTORE_PREFER_REST`. If not specified, the
+   * SDK will use the value specified in environment variable `FIRESTORE_PREFER_REST=true|false`. And if the
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply dev/src/index.ts]: IMO this isn't a great place to go into detail about the behavior of the environment variable, but after looking around, I don't think we have a better place to document environment variables used by the SDK. PTAL at the changes here and let me know what you think.

File: dev/src/util.ts [hunk 4]
```diff
@@ -217,3 +217,29 @@ export function wrapError(err: Error, stack: string): Error {
...
+
+/**
+ * Returns a tuple indicating if FIRESTORE_PREFER_REST is set to a valid value,
+ * and if so, whether the value enables preferRest or disables preferRest.
+ *
+ * @return A tuple where the first value indicates if the environment variable is set to a valid value,
+ * and the second value indicates if the environment variable enables preferRest or disables preferRest.
+ */
+export function tryGetPreferRestEnvironmentVariable():
+  | [false, undefined]
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply dev/src/util.ts]: Hmm. This design is an artifact of an older implementation. It no longer makes sense. Thanks for calling it out.

File: dev/test/index.ts [hunk 5]
```diff
@@ -639,6 +639,58 @@ describe('instantiation', () => {
...
+        delete process.env.FIRESTORE_PREFER_REST;
+      } else {
+        process.env.FIRESTORE_PREFER_REST = originalValue;
+      }
+    });
+
+    it('preferRest is disabled (falsy) by default', async () => {
+      delete process.env.FIRESTORE_PREFER_REST;
+      const firestore = new Firestore.Firestore({});
+      expect(!!firestore['_settings'].preferRest).to.be.false;
```
[the remark this answers is not in the document]
Comment (MarkDuckworth) [reply dev/test/index.ts:664]: Added a comment to clarify. Curious what you think about alternatively initializing this setting value to `false` (instead of leaving it as undefined) in the case where the setting is not set by the consumer?

File: dev/test/util.ts (hunk 2 above)
Comment (MarkDuckworth) [reply dev/test/util.ts:83]: added. thanks.

File: dev/test/util.ts (hunk 1 above)
Comment (MarkDuckworth) [reply dev/test/util.ts]: let me know if the new test name clarifies it for you. thanks

File: dev/src/util.ts [hunk 6]
```diff
@@ -219,27 +219,28 @@ export function wrapError(err: Error, stack: string): Error {
...
 
 /**
- * Returns a tuple indicating if FIRESTORE_PREFER_REST is set to a valid value,
- * and if so, whether the value enables preferRest or disables preferRest.
+ * Parses the value of the environment variable FIRESTORE_PREFER_REST, and returns
+ * a value indicating if the environment variable enables or disables preferRest.
  *
- * @return A tuple where the first value indicates if the environment variable is set to a valid value,
- * and the second value indicates if the environment variable enables preferRest or disables preferRest.
+ * This method will warn to the console if the environment variable is set to an unsupported value.
```
Comment (dconeybe) [inline dev/src/util.ts:225]: ultranit: "method" -> "function" (a "method" is a member of a class, which this function is not).

File: dev/src/util.ts [hunk 7]
```diff
@@ -219,27 +219,28 @@ export function wrapError(err: Error, stack: string): Error {
...
+ *
+ * @return {boolean | undefined} `true` if the environment variable enables `preferRest`,
```
Comment (dconeybe) [inline dev/src/util.ts:227]: nit: I'd recommend to remove the definition of the return type from the documentation since it's already specified formally in the function's return type declaration. By repeating it here you risk it getting out of sync with the actual return type, which is confusing.

File: dev/src/util.ts [hunk 8]
```diff
@@ -219,27 +219,28 @@ export function wrapError(err: Error, stack: string): Error {
...
+ * `false` if the environment variable disable `preferRest`, or `undefined` if the
```
Comment (dconeybe) [inline dev/src/util.ts:228]: ultranit: "disable" -> "disables"

File: dev/test/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dconeybe) [reply dev/test/index.ts:664]: IMO leaving it as `undefined` is preferred because it more clearly conveys that the setting is optional. Thanks for adding the comment; that is helpful (to me!).

File: dev/test/util.ts [hunk 9]
```diff
@@ -58,51 +58,50 @@ describe('isPlainObject()', () => {
 
     it('reads true', async () => {
       process.env.FIRESTORE_PREFER_REST = 'true';
-      const [prevSet, prevValue] = tryGetPreferRestEnvironmentVariable();
-      expect(prevSet).to.be.true;
+      const prevValue = tryGetPreferRestEnvironmentVariable();
```
Comment (dconeybe) [inline dev/test/util.ts:61]: nit: Can you rename `prevValue` to avoid the abbreviation? I don't know what "prev" is an abbreviation for. Is it "previous"? Oh, is it an acronym for "prefer rest environment variable"? Consider a name like "returnValue", "result" or "actual", or even just "value". Even better, get rid of the local variable and merge it into the next line, like this:

```ts
expect(tryGetPreferRestEnvironmentVariable()).to.be.true;
```

I actually find that to be the most readable because it "reads like a story" from left to right.

In general, non-standard abbreviations detract from readability. Avoid them.

Applies here and elsewhere in this file.

File: dev/test/util.ts (hunk 1 above)
Comment (dconeybe) [reply dev/test/util.ts]: It is perfect. Thank you.

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: dev/src/util.ts (hunk 7 above)
Comment (MarkDuckworth) [reply dev/src/util.ts:227]: Good way to look at it.

File: dev/src/util.ts [hunk 10]
```diff
@@ -222,10 +222,10 @@ export function wrapError(err: Error, stack: string): Error {
  * Parses the value of the environment variable FIRESTORE_PREFER_REST, and returns
  * a value indicating if the environment variable enables or disables preferRest.
  *
- * This method will warn to the console if the environment variable is set to an unsupported value.
+ * This function will warn to the console if the environment variable is set to an unsupported value.
```
Comment (dconeybe) [inline dev/src/util.ts:225]: nit: wrap at 80 characters

**Review by dconeybe — CHANGES_REQUESTED** [review]

File: dev/src/util.ts (hunk 10 above)
Comment (MarkDuckworth) [reply dev/src/util.ts]: FWIW, (gts)[https://github.com/google/gts] enforces Google's typescript style guide and does not enforce that column limit on comments and some other line types. It's unclear if that's because there's an issue or the 80-column limit does not actually apply.

**Review by dconeybe — APPROVED** [review]
